R/recover.weaker.R

```diff
@@ -2,172 +2,644 @@
   }
-  return(this.found)
+
+  return(list(intensity = sc, label = miu))
 }
 
+#' Compute the rectangle around recovered features given that enough peaks are present.
+#' @description
+#'
+#' @param mz Mz value of the feature.
+#' @param peak Peak around which to detect the new feature.
+#' @param valleys Valley points to compute the boundary region.
+#' @param features tibble Tibble with `labels` and `intensities` column.
+#' @param aver_diff float Average retention time difference.
+#' @param times vector Raw retention time values from raw data file.
+#' @param delta_rt vector Differences between consecutive retention time values (diff(times)).
+#' @return list Triplet of mz, label and intensity for the feature.
+compute_curr_rec_with_enough_peaks <- function(mz,
+                                               peak,
+                                               valleys,
+                                               features,
+                                               aver_diff,
+                                               times,
+                                               delta_rt) {
+  # same filtering of peaks as in compute_pks_vlyws and as above
+  boundaries <- compute_boundaries(valleys, peak)
+
+  subset <- features |>
+    dplyr::filter(labels >= boundaries$lower & labels <= boundaries$upper)
+
+  if (nrow(subset) == 1) {
+    intensity <- subset$intensities * aver_diff
+    label <- subset$labels
+  } else if (nrow(subset) >= 10) {
+    res <- compute_mu_sc(subset, aver_diff)
+    intensity <- res$intensity
+    label <- res$label
+  } else {
+    intensity <- interpol.area(
+      subset$labels,
+      subset$intensities,
+      times,
+      delta_rt
+    )
+    label <- median(subset$labels)
+  }
+
+  return(c(mz, label, intensity))
+}
+
+#' Compute bounds of area using given peak and mass valley points.
+#' @description
+#' The lower bound is the mass of the valley the closest but smaller than peak
+#' and the upper bound is the mass of the valley the closest but higher than
+#' the peak.
+#' @param valley_points vector values of valley points defining clusters.
+#' @param peak double Value of the peak for which to get the valley bounds.
+#' @return Returns a list object with the following objects in it:
+#' \itemize{
+#'   \item lower - double - The value of the lower bound valley point.
+#'   \item upper - double - The value of the upper bound valley point.
+#' }
 #' @export
-get_raw_features_in_mzrange <- function(data_table, aligned.ftrs, breaks, i, custom.mz.tol) {
-  this.found <- get_mzrange_bound_indices(aligned.ftrs, data_table$mz, breaks, i, custom.mz.tol)
-  this.sel <- (breaks[this.found[1]] + 1):breaks[this.found[2]]
-  features <- data_table |> dplyr::slice(this.sel)
-  return(features)
+compute_boundaries <- function(valley_points, peak) {
+  lower <- max(valley_points[valley_points < peak])
+  upper <- min(valley_points[valley_points > peak])
```

Comment:
This pattern seems to be occurring several times. Maybe we could extract it to a function that returns a vector `c(lower, upper)`, akin R's `range()`?

Reply:
After having another look I realised I was probably wrong. In the other case it computed indices, not values.